chart/templates/hook-upgrade-crds.yaml

```diff
@@ -0,0 +1,116 @@
+{{- /*
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: BUSL-1.1
+*/ -}}
+
+{{- if .Values.hooks.upgradeCRDs.enabled -}}
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ template "vso.chart.fullname" . }}-upgrade-crds
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app.kubernetes.io/component: controller-manager
+{{ include "vso.chart.labels" . | indent 4 }}
+  annotations:
+    helm.sh/hook: pre-upgrade
+    helm.sh/hook-delete-policy: "hook-succeeded,before-hook-creation"
+    helm.sh/hook-weight: "1"
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: {{ template "vso.chart.fullname" . }}-upgrade-crds
+  labels:
+    app.kubernetes.io/component: rbac
+{{ include "vso.chart.labels" . | indent 4 }}
+  annotations:
+    helm.sh/hook: pre-upgrade
+    helm.sh/hook-delete-policy: "hook-succeeded,before-hook-creation"
+    helm.sh/hook-weight: "2"
+rules:
+  - apiGroups:
+    - apiextensions.k8s.io
+    resources:
+    - customresourcedefinitions
+    verbs:
+    - create
+    - delete
+    - get
+    - list
+    - patch
+    - update
```

Comment:
Random thought: is there a way to limit this to just our CRDs? At least for the write verbs.

Reply:
We would probably have to `resourceNames` statically to get something like what you are after here. That might be a bit brittle since we would have to update the set of names whenever we add new CRDs. Since this ClusterRole is created and destroyed for the lifetime of the hook, I think we are probably pretty safe here.